driver/api/impl/impl.cpp

```diff
@@ -0,0 +1,1284 @@
+                        statement.getParent().getAttrAs<SQLUINTEGER>(SQL_ATTR_METADATA_ID, SQL_FALSE)
+                    ),
+                    out_value, out_value_length
+                );
+
+            CASE_FALLTHROUGH(SQL_ATTR_NOSCAN)
+                return fillOutputPOD<SQLULEN>(
+                    statement.getAttrAs<SQLULEN>(SQL_ATTR_NOSCAN, SQL_NOSCAN_OFF),
+                    out_value, out_value_length
+                );
+
+            CASE_NUM(SQL_ATTR_QUERY_TIMEOUT, SQLULEN, 0);
+            CASE_NUM(SQL_ATTR_RETRIEVE_DATA, SQLULEN, SQL_RD_ON);
+            CASE_NUM(SQL_ATTR_ROW_NUMBER, SQLULEN, statement.getCurrentRowNum());
+            CASE_NUM(SQL_ATTR_USE_BOOKMARKS, SQLULEN, SQL_UB_OFF);
+
+            case SQL_ATTR_FETCH_BOOKMARK_PTR:
+            case SQL_ATTR_KEYSET_SIZE:
+            case SQL_ATTR_SIMULATE_CURSOR:
+            default:
+                LOG("GetStmtAttr: Unsupported attribute " << attribute);
+                //throw std::runtime_error("Unsupported statement attribute. " + std::to_string(attribute));
+                return SQL_ERROR;
+        }
+
+        return SQL_SUCCESS;
+    };
+
+    return CALL_WITH_TYPED_HANDLE(SQL_HANDLE_STMT, statement_handle, func);
+}
+
+SQLRETURN GetDiagRec(
+    SQLSMALLINT handle_type,
+    SQLHANDLE handle,
+    SQLSMALLINT record_number,
+    SQLTCHAR * out_sqlstate,
+    SQLINTEGER * out_native_error_code,
+    SQLTCHAR * out_message,
+    SQLSMALLINT out_message_max_size,
+    SQLSMALLINT * out_message_size
+) noexcept {
+    auto func = [&] (auto & object) -> SQLRETURN {
+        if (record_number < 1 || out_message_max_size < 0)
+            return SQL_ERROR;
+
+        if (record_number > object.getDiagStatusCount())
+            return SQL_NO_DATA;
+
+        const auto & record = object.getDiagStatus(record_number);
+
+        if (out_sqlstate) {
+            std::size_t size = 6;
+            std::size_t written = 0;
+            fillOutputString<SQLTCHAR>(record.template getAttrAs<std::string>(SQL_DIAG_SQLSTATE), out_sqlstate, size, &written, false);
+        }
+
+        if (out_native_error_code != nullptr) {
+            *out_native_error_code = record.template getAttrAs<SQLINTEGER>(SQL_DIAG_NATIVE);
+        }
+
+        return fillOutputString<SQLTCHAR>(record.template getAttrAs<std::string>(SQL_DIAG_MESSAGE_TEXT), out_message, out_message_max_size, out_message_size, false);
+    };
+
+    return CALL_WITH_TYPED_HANDLE_SKIP_DIAG(handle_type, handle, func);
+}
+
+SQLRETURN GetDiagField(
+    SQLSMALLINT handle_type,
+    SQLHANDLE handle,
+    SQLSMALLINT record_number,
+    SQLSMALLINT field_id,
+    SQLPOINTER out_message,
+    SQLSMALLINT out_message_max_size,
+    SQLSMALLINT * out_message_size
+) noexcept {
+    auto func = [&] (auto & object) -> SQLRETURN {
+        // Exit with error if the requested field is relevant only to statements.
+        if (getObjectHandleType<decltype(object)>() != SQL_HANDLE_STMT) {
```

Comment:
IMO, that should be:
```
if (getObjectHandleType<std::decay_t<decltype(object)>>() != SQL_HANDLE_STMT) {
```
Also, I think it is worth removing function definition of `template <typename T> constexpr int getObjectHandleType();` [here](https://github.com/ClickHouse/clickhouse-odbc/blob/08b252b93fc771fab607fb973443543479b3d972/driver/utils/utils.h#L73).

Reply:
Done.